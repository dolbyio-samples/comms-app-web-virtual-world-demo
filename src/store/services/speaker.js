import VoxeetSDK from "@voxeet/voxeet-web-sdk";

export const isSpeaking = async (participant, cb) => {
  try {
    await VoxeetSDK.conference.isSpeaking(participant, cb);
  } catch (error) {
    return { error };
  }
};

export const audioLevel = async (participant, cb) => {
  try {
    await VoxeetSDK.conference.audioLevel(participant, cb);
  } catch (error) {
    return { error };
  }
};
