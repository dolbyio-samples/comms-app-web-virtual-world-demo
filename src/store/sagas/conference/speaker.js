import {
  take,
  fork,
  cancel,
  cancelled,
  put,
  delay,
  race,
  call,
  all,
} from "redux-saga/effects";
import { isSpeaking as isSpeakingService } from "../../services/speaker";

import { stopBackgroundTask } from "../../actions/conference";
import { conference, application } from "../../effects";
import { displayMessage } from "./message";

//let participantId;
export function* bgSync(participant) {
  let isSpeaking = false;
  let previousStatus = false;

  const handleSpeakingStatus = (status) => {
    isSpeaking = status;
  };

  try {
    while (true) {
      try {
        // effects will get executed in parallel
        yield all([call(isSpeakingService, participant, handleSpeakingStatus)]);

        if (isSpeaking !== previousStatus) {
          previousStatus = isSpeaking;
          yield put(
            conference.updateParticipant({
              id: participant.id,
              entityType: "participants",
              data: { isSpeaking },
            })
          );
        }
      } catch (error) {
        yield call(displayMessage, {
          type: "error",
          message: "could not retrieve speaking status",
        });
      }

      yield delay(300);
    }
  } finally {
    if (yield cancelled()) {
      // console.log("task cancelled");
      yield put(stopBackgroundTask());
    }
  }
}

export function* watchParticipant({ payload: { data, id } }) {
  const bgSyncTask = yield fork(bgSync, data[id]);

  while (true) {
    // console.log("waiting for participant updated");

    const { update, remove, ends } = yield race({
      update: take(conference.updateParticipant),
      remove: take(conference.removeParticipant),
      ends: take([
        conference.ended,
        conference.left,
        application.leaveConference,
      ]),
    });

    if (ends) {
      yield cancel(bgSyncTask);
      break;
    }

    if (remove) {
      if (id === remove.payload.id) {
        yield cancel(bgSyncTask);
        break;
      }
    }

    if (update) {
      if (
        id === update.payload.id &&
        update.payload.data.status &&
        update.payload.data.status === "Left"
      ) {
        yield cancel(bgSyncTask);
        break;
      }
    }
  }
  // console.log("exiting");
}
