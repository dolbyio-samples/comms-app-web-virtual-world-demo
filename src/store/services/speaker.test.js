import { describe, it, expect, vi } from "vitest";
import { audioLevel, isSpeaking } from "./speaker";

const mockIsSpeaking = vi.fn();
const mockAudioLevel = vi.fn();

vi.mock("@voxeet/voxeet-web-sdk", () => ({
  default: {
    get conference() {
      return { isSpeaking: mockIsSpeaking, audioLevel: mockAudioLevel };
    },
  },
}));

describe("speaker services test suite", () => {
  it("should succeed", async () => {
    // 1- arrange
    const participant = { name: "toto" };
    const callback = () => {};

    // act
    const result = await isSpeaking(participant, callback);

    // assert
    expect(mockIsSpeaking).toHaveBeenCalled();
    expect(mockIsSpeaking).toHaveBeenCalledWith(participant, callback);
    expect(result).toBe(undefined);
  });

  it("should fail and return an error", async () => {
    // arrange
    const participant = { name: "toto" };
    const callback = () => {};
    mockIsSpeaking.mockImplementation(() => {
      throw new Error("error");
    });

    // act
    const { error } = await isSpeaking(participant, callback);

    // assert
    expect(error).not.toBe(null);
    expect(error).toMatchSnapshot();
    expect(mockIsSpeaking).toHaveBeenCalled();
    expect(mockIsSpeaking).toHaveBeenCalledWith(participant, callback);
  });

  it("should succeed", async () => {
    // arrange
    const participant = { name: "toto" };
    const callback = () => {};

    // act
    const result = await audioLevel(participant, callback);

    // assert
    expect(mockAudioLevel).toHaveBeenCalled();
    expect(mockAudioLevel).toHaveBeenCalledWith(participant, callback);
    expect(result).toBe(undefined);
  });

  it("should fail and return an error", async () => {
    // arrange
    const participant = { name: "toto" };
    const callback = () => {};
    mockAudioLevel.mockImplementation(() => {
      throw new Error("error");
    });

    // act
    const { error } = await audioLevel(participant, callback);

    // assert
    expect(error).not.toBe(null);
    expect(error).toMatchSnapshot();
    expect(mockAudioLevel).toHaveBeenCalled();
    expect(mockAudioLevel).toHaveBeenCalledWith(participant, callback);
  });
});
