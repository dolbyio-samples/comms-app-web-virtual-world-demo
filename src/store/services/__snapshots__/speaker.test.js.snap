// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`speaker services test suite > should fail and return an error 1`] = `[Error: error]`;

exports[`speaker services test suite > should fail and return an error 2`] = `[Error: error]`;
